### PR TITLE
exit with status code 1 when the report is not empty

### PR DESCRIPTION
**Pull request by AndreaCrotti**
This is particularly useful for things like CI where having some dead
code is supposed to make the build fail.

### CODE REVIEW

**Comment by AndreaCrotti** [comment]
I just tried it with two projects with and without an empty report and it seems to work as expected

```
~/src/payments-reconciler $ clojure -Acarve --opts '{:paths ["src" "test"] :report {:format :text}}'
...
test/payments_reconciler/test_utils.clj:178:1 payments-reconciler.test-utils/kaocha
test/payments_reconciler/test_utils.clj:182:1 payments-reconciler.test-utils/kaocha-all
~/src/payments-reconciler $ echo $?
1
~/src/payments-reconciler $ cd ../
~/src $ cd ../projects/forks/carve/
~/projects/forks/carve $ clojure -Acarve --opts '{:paths ["src" "test"] :report {:format :text}}'
~/projects/forks/carve $ echo $?
0
```

**Comment by AndreaCrotti** [comment]
Tests fail, I guess I can just add a with-redefs or similar to avoid making it exit during testing

**Comment by borkdude** [comment]
@AndreaCrotti I'd rather not use with-redefs, but test for real if the program returns exit code 1 in case of unused vars. See clj-kondo how this is done. There is `-main` which calls `main`. `main` returns the exit code and this is tested in the unit tests. `-main` then simply does `(System/exit (main ...))`.

**Comment by AndreaCrotti** [comment]
Yes ok sure, it also breaks tests as it is now so with your suggestion that would fix it.

File: src/carve/main.clj [hunk 1]
```diff
@@ -43,12 +43,18 @@
...
 
-(defn -main [& [flag opts & _args]]
+(defn main
+  [& [flag opts & _args]]
   (when-not (= "--opts" flag)
     (throw (ex-info (str "Unrecognized option: " flag) {:flag flag})))
   (let [opts (edn/read-string opts)
         _ (validate-opts! opts)
         report (impl/run! opts)]
     (when-let [r (:report opts)]
```
Comment (borkdude) [inline src/carve/main.clj:53]: I guess `main` can potentially return nil. I should it should return `0` instead of `nil` as the default exit code.

Comment (AndreaCrotti) [reply src/carve/main.clj:53]: Well I just wanted to change it for the report mode, I could change the rest as well and default to 0 but I think it might be wrong in way.
Or at least we would have to decide what to return in the cases where it's not a report, returning always `0` might not be correct imho, `nil` just says that it wasn't a report at least.

Comment (borkdude) [reply src/carve/main.clj:53]: What does `(System/exit nil)` do?

Comment (AndreaCrotti) [reply src/carve/main.clj:53]: Ah good question, it blows up 😄 
I only tried report otherwise I would have noticed.
Yeah well I can just add a `some->` maybe if that's ok, otherwise I can also just default to 0 what do you prefer @borkdude ?

Comment (AndreaCrotti) [reply src/carve/main.clj:53]: It's the same thing I guess since if we don't call System/exit directly the exit status code is 0 anyway.

Comment (borkdude) [reply src/carve/main.clj:53]: I think it's cleaner to have `main` always have it return a number, either 1 or 0. Just use `(or (...) 0)` I'd say.

Comment (AndreaCrotti) [reply src/carve/main.clj:53]: Ok cool

File: src/carve/main.clj [hunk 2]
```diff
@@ -43,12 +43,18 @@
...
-    (when-let [r (:report opts)]
-      (let [format (:format r)]
-        (impl/print-report report format)))))
+    (when format
+      (impl/print-report report format))
+    (if report 1 0)))
+
+(defn -main
+  [& options]
+  (System/exit (or (apply main options) 0)))
```
Comment (borkdude) [inline src/carve/main.clj:60]: I prefer `main` to always return a number so `-main` is nothing more than just the `System/exit` + call to `main`.

File: src/carve/main.clj [hunk 3]
```diff
@@ -43,12 +43,18 @@
...
   (let [opts (edn/read-string opts)
         _ (validate-opts! opts)
+        format (-> opts :report :format)
         report (impl/run! opts)]
-    (when-let [r (:report opts)]
-      (let [format (:format r)]
-        (impl/print-report report format)))))
+    (when format
+      (impl/print-report report format))
+    (if report 1 0)))
```
Comment (borkdude) [inline src/carve/main.clj:56]: You are relying on `report` being something truthy. It it ever `nil`? Shouldn't you return `0` if the report is empy?

Comment (borkdude) [reply src/carve/main.clj:56]: I think it would also be good to add unit tests for testing `0` or `1` for a zero and a positive amount of findings.

Comment (AndreaCrotti) [reply src/carve/main.clj]: Yes sorry you're right, I fixed it and added a couple of tests.

File: src/carve/main.clj (hunk 2 above)
Comment (borkdude) [reply src/carve/main.clj:60]: Now we can probably remove the `or` in `-main` since `main` now always returns a number.

Comment (borkdude) [reply src/carve/main.clj:60]: You marked this as resolved:

> Now we can probably remove the or in -main since main now always returns a number.

But this wasn't resolved?

Comment (AndreaCrotti) [reply src/carve/main.clj]: Done!

**Comment by borkdude** [comment]
Thanks for the PR!
